src/tree/updater_gpu_hist.cu

```diff
@@ -1264,277 +1469,46 @@ class GPUHistMakerSpecialised{
-      shard->node_sum_gradients[tree[candidate.nid].RightChild()] = candidate.split.right_sum;
-    }
-  }
 
   void UpdateTree(HostDeviceVector<GradientPair>* gpair, DMatrix* p_fmat,
                   RegTree* p_tree) {
-    auto& tree = *p_tree;
-
     monitor_.StartCuda("InitData");
     this->InitData(gpair, p_fmat);
     monitor_.StopCuda("InitData");
-    monitor_.StartCuda("InitRoot");
-    this->InitRoot(p_tree);
-    monitor_.StopCuda("InitRoot");
-
-    auto timestamp = qexpand_->size();
-    auto num_leaves = 1;
 
-    while (!qexpand_->empty()) {
-      ExpandEntry candidate = qexpand_->top();
-      qexpand_->pop();
-      if (!candidate.IsValid(param_, num_leaves)) {
-        continue;
-      }
+    std::vector<RegTree> dummy_trees(shards_.size());
+    for (auto& tree : dummy_trees) {
+      tree = *p_tree;
+    }
+    gpair->Reshard(dist_);
 
-      this->ApplySplit(candidate, p_tree);
-      num_leaves++;
+    // Launch one thread for each device "shard" containing a subset of rows.
+    // Threads will cooperatively build the tree, synchronising over histograms.
+    // Each thread will redundantly
+    // build its own copy of the tree
+    dh::ExecuteIndexShards(
+        &shards_,
+        [&](int idx, std::unique_ptr<DeviceShard<GradientSumT>>& shard) {
+          // Only the first shard will actually modify the tree
+          RegTree* tree = idx == 0 ? p_tree : &dummy_trees.at(idx);
+          shard->UpdateTree(gpair, p_fmat, tree,
+                          &reducer_);
+        });
 
-      int left_child_nidx = tree[candidate.nid].LeftChild();
-      int right_child_nidx = tree[candidate.nid].RightChild();
-      // Only create child entries if needed
-      if (ExpandEntry::ChildIsValid(param_, tree.GetDepth(left_child_nidx),
-                                    num_leaves)) {
-        monitor_.StartCuda("UpdatePosition");
-        this->UpdatePosition(candidate, p_tree);
-        monitor_.StopCuda("UpdatePosition");
-
-        monitor_.StartCuda("BuildHist");
-        this->BuildHistLeftRight(candidate.nid, left_child_nidx,
-                                 right_child_nidx);
-        monitor_.StopCuda("BuildHist");
-
-        monitor_.StartCuda("EvaluateSplits");
-        auto splits =
-            this->EvaluateSplits({left_child_nidx, right_child_nidx}, p_tree);
-        qexpand_->push(ExpandEntry(left_child_nidx,
-                                   tree.GetDepth(left_child_nidx), splits.at(0),
-                                   timestamp++));
-        qexpand_->push(ExpandEntry(right_child_nidx,
-                                   tree.GetDepth(right_child_nidx),
-                                   splits.at(1), timestamp++));
-        monitor_.StopCuda("EvaluateSplits");
-      }
+    // Check the shards produced identical trees
+    for (auto i = 1ull; i < dummy_trees.size(); i++) {
+      CHECK(*p_tree == dummy_trees.at(i));
     }
```

Comment:
Depending on the tree size and the number of GPUs, this may be quite an overhead. Also note that it runs serially for any number of GPUs. It also doesn't check that the trees are the same in a distributed setting.

I wonder if it is possible to put this check behind a parameter.

Also, is there a python test that runs with multiple GPUs in a single process?

Reply:
I added a parameter for this and extended the check to distributed mode using rabit broadcast. The python multi-GPU tests will cover the case of a single process with multiple GPUs.